### PR TITLE
Improve CUDA out-of-memory error message

### PR DESCRIPTION
**Pull request by colesbury**
```
The new error message now looks like (from Python):

  RuntimeError: CUDA out of memory. Tried to allocate 16.00 GiB (GPU 0; 11.93 GiB total capacity; 4.00 GiB already allocated; 7.33 GiB free; 179.00 KiB cached)

Summary of terms:

  "total capacity": total global memory on GPU
  "already allocated": memory allocated by the program using the
                       caching allocator
  "free": free memory as reported by the CUDA API
  "cached": memory held by the allocator but not used by the program
 
  The "allocated" amount  does not include memory allocated outside
  of the caching allocator, such as memory allocated by other programs
  or memory held by the driver.
 
  The sum of "allocated" + "free" + "cached" may be less than the
  total capacity due to memory held by the driver and usage by other
  programs.

  Note that at this point cuda_malloc_retry has already returned all
  possible "cached" memory to the driver. The only remaining "cached"
  memory is split from a larger block that is partially in-use.
```

This also fixes an issue where on out-of-memory could cause an unrelated subsequent CUDA kernel launch to fail because `cudaGetLastError()` was not cleared.

### CODE REVIEW

File: aten/src/THC/THCCachingAllocator.cpp [hunk 1]
```diff
@@ -148,28 +167,23 @@ struct THCCachingAllocator
...
+    process_events();
 
     size = round_size(size);
     bool small = size <= kSmallAlloc;
 
     DeviceStats &stats = get_stats_for_device(device);
 
     Block search_key(device, stream, size);
-    auto& free_blocks = small ? large_blocks : small_blocks;
+    auto& free_blocks = small ? small_blocks : large_blocks;
```
Comment (pietern) [inline aten/src/THC/THCCachingAllocator.cpp]: Could this have been the culprit?

Comment (pietern) [reply aten/src/THC/THCCachingAllocator.cpp]: If small allocations are routed to large blocks, they can't be freed anymore and will cause more fragmentation then if they are always routed to the small blocks.

Comment (colesbury) [reply aten/src/THC/THCCachingAllocator.cpp]: No. The names were swapped, but the logic was consistent.

File: aten/src/THC/THCCachingAllocator.cpp [hunk 2]
```diff
@@ -181,9 +195,36 @@ struct THCCachingAllocator
...
       size_t alloc_size = small ? kSmallAlloc : size;
-      err = cuda_malloc_retry(device, &ptr, alloc_size);
+      cudaError_t err = cuda_malloc_retry(device, &ptr, alloc_size);
       if (err != cudaSuccess) {
-        return err;
+        if (err == cudaErrorMemoryAllocation) {
+          cudaGetLastError();  // clear CUDA error
+
+          cudaDeviceProp prop;
+          AT_CUDA_CHECK(cudaGetDeviceProperties(&prop, device));
```
Comment (apaszke) [inline aten/src/THC/THCCachingAllocator.cpp]: This is a super slow call, and we have those cached in ATen context, so it would be good to use that.

**Review by apaszke — APPROVED** [review]

File: aten/src/THC/THCCachingAllocator.cpp (hunk 1 above)
Comment (apaszke) [reply aten/src/THC/THCCachingAllocator.cpp]: Hmmm how is it possible that this is equivalent? I can’t see where’s the other swap that would cancel this one out.

Comment (colesbury) [reply aten/src/THC/THCCachingAllocator.cpp]: In `free_block()`. Other than insertion and removal, the `small_blocks` and `large_blocks` block lists are treated identically.

(Small allocations are rounded differently from large allocations, but that's dependent on the variable `small` not the choice of block list)

File: aten/src/THC/THCCachingAllocator.cpp (hunk 2 above)
Comment (colesbury) [reply aten/src/THC/THCCachingAllocator.cpp]: I switched to cudaMemGetInfo to be able to report the free memory as well

**Pull request closed without merging**